### PR TITLE
Phoenix Chat

### PR DESCRIPTION
**Pull request by nelsonic**
Good Morning @katbow! ☀️ 
This PR adds a "Spike" of building a Chat System in Phoenix using OTP Channels.
Checkout the branch and try it out! 👍 

@iteles _please review & merge_ if ok. thanks!

### CODE REVIEW

**Comment by jacktuck** [comment]
So no more serverless ? :( Ever seen https://azure.microsoft.com/en-gb/pricing/details/service-bus/ ?

**Review by iteles — APPROVED** [review]
Wow, nice!

Might be good to add a line to the readme to explain why we chose to move away from Lambda too for those who might be curious 👍 

Amazing work, thanks for putting in the time!
Inês

**Comment by nelsonic** [comment]
Hi @jacktuck, we still _like_ "serverless", especially for infrequent or batch processing but we noted some pretty meaningful bottlenecks when using Serverless for an entire application. 
see: https://github.com/numo-labs where we Open-Sourced _most_ of the components in the stack.

Ultimately we tried _everything_ to reduce latency (_improve response times for user requests_) and we found ourselves "_fighting_" with the "system" too much so eventually we decided to "_go back to basics_". 

> If we can build a chat app that is ***much faster*** and costs less to operate by using Phoenix,
doesn't that sound like "_progress_"...? 💭 

How much have you researched the most successful instant messaging platforms/apps?
(_not trying to be condescending, just curious if this is something you're interested in..._)

If you were building "Slack" or "Messenger" _from scratch_ today, 
would you use a 3rd party to deliver your messages? 
Or would you consider message routing and delivery to be the "_core_" of your app?

Personally, if I were considering building "chat" from _scratch_ (_which we are...!_) 
I would look at the most popular (_existing_) messaging services, 
study their "stack" to _understand **why**_ they selected a specific technology
and go and _learn_ that technology. 

In the case of "Chat" (_instant messaging_) the technology of choice is Erlang. 
And since we want _humans_ to be able to read/maintain/modify the code,
we decided to use Elixir which _compiles_ to run on the Erlang VM.

Apologies for the "long" answer, this probably deserves a "blog post"... 😉 

Anyway, the "Serverless" version lives on: https://github.com/dwyl/aws-lambda-iot-chat for anyone who is interested in taking it further.

Our plan for _this_ repo is to flesh out the features and make it 100% platform/infrastructure agnostic so people can run it _anywhere_ at the _lowest possible cost_.

# tl;dr

The `promise` of "serverless" is not being fulfilled. 
The differences between service providers/vendors is _increasing_ as they each strive to achieve dominance in the market. 
It's like the 1990's ["browser wars"](https://en.wikipedia.org/wiki/Browser_wars) all over again, but on the back-end.
There is no standardisation so vendors (AWS, Google, Microsoft, etc) are competing to add features in _slightly_ different ways ... 

Yes, there are frameworks like Serverless or Apex that are attempting to "smooth over" the differences between vendors and present an abstraction so developers don't have to think about "_implementation detail_" ... Sadly, we have discovered (_through experience_) that after you've done a _basic_ "Serverless" app you _quickly_ run into some significant performance (_latency_) bottlenecks which impact the UX of your web app. 

> Would you write a "Chat" (_instant messaging_) application using Serverless? 
> The ***core*** of "real-time" is message delivery/broadcasting as discussed in https://github.com/dwyl/chat/issues/26
> Message delivery is _not_ currently being addressed by any of the cloud providers.
> Would you write a video chat application using "Serverless" or use WebRTC?
> We would use the "_best tool_" for the job. in our case a combination of OTP which upgrades to WebRTC when both endpoints support it.

I think the benefits of running on the Erlang VM (BEAM) _greatly_ outweigh the "cost" of paying for a single server instance and having a little bit of idle time. 
e.g: $5/month for a server that responds in 50ms instead of taking 800-2000ms to "warm up". 

We are still going to use AWS Lambda (_or equivalent_) for specific self-contained functions, 
but for chat where latency _matters_ we want to be as close to the protocol (_in this case **OTP**_) as possible to give the best UX to the people _using_ the application.

Yes, we have seen the "Azure Service Bus" it's a _good_ solution for people building "micro services".
We have _used_ the AWS equivalent https://aws.amazon.com/sqs for our apps in the past and it works well. But while the benefits of "not having to think about infrastructure" are nice in _theory_, each "_hop_" in your stack adds time for a request to be processed which means people are waiting ... 
under the hood they are using AMQP https://en.wikipedia.org/wiki/Advanced_Message_Queuing_Protocol which is a _good_ standard.
But Erlang gives you this kind of message passing/queuing functionality "built in"!
So _by definition_ an application build in Erlang (_or Elixir which compiles to BEAM_) will run faster than one which breaks up its components and uses a 3rd party messaging layer. 

As an _individual_ developer I would recommend using "Serverless" to get started building an app if the objective is to get an "MVP" to end-users as soon as possible.
But anyone looking for more _understanding_ of their "_Stack_" and the knowledge to "_squeeze_" the most performance out of their units of compute power 
should _seriously_ consider Erlang |> Elixir |> Phoenix _instead_ of "Serverless".

I _highly_ recommend reading: 
http://www.phoenixframework.org/blog/the-road-to-2-million-websocket-connections

These days people have a "Phobia" around _owning_ (_and running_) Hardware, 
this has been ***fueled*** by "Cloud" vendors who _obviously_ stand to benefit from nobody owning/operating hardware and handing over their cash to the IaaS/PaaS supplier.
And the "Serverless" mindset is a perpetuation of this trend.
What is the end result? we become _dependent_ on the supplier; [locked-in](https://en.wikipedia.org/wiki/Vendor_lock-in). 

Again, I think it's worth splitting this answer out into a "blog post" to cover each topic in a bit more detail. maybe I will find the time _some day_... 😉

**Comment by jacktuck** [comment]
Thanks for taking the time to reply!

I have some thoughts on some of the topic's you've brought up :)

>http://www.phoenixframework.org/blog/the-road-to-2-million-websocket-connections

I've given that a read and 2m connections on one server is _very_ impressive but they haven't benchmarked what other implementations would have achieved on the same hardware. Also the hardware they used comes at an enormous cost at 3.84/h
(3.84 * 24) * 30.44 = 2805.3504 😱 

>How much have you researched the most successful instant messaging platforms/apps?

A fair bit. And you're right, I know of none that use serverless.

I think there are a few motivations for them to not use serverless
- Not being able to offer an enterprise package where the product is run 'behind the firewall'.
- Vendor lock-in
- Lack of expertise on the team
- Lack  observability

But saying that _X is more common, therefore X is the right approach_ is an appeal to common practice fallacy.

**I think I read that you were building microservices?**
That is what our team do, but i think when you're building microservices what you're really starting to go towards is a decoupled architecture and nothing offers better decoupling than FaaS

>I think the benefits of running on the Erlang VM (BEAM) greatly outweigh the "cost" of paying for a single server instance and having a little bit of idle time.
e.g: $5/month for a server that responds in 50ms instead of taking 800-2000ms to "warm up".

Fair point, but somewhat hyperbolic, there is only a cold start cost for newly created containers, once you have constant traffic you should not encounter that anymore.

>These days people have a "Phobia" around owning (and running) Hardware,
this has been fueled by "Cloud" vendors who obviously stand to benefit from nobody owning/operating hardware and handing over their cash to the IaaS/PaaS supplier.
And the "Serverless" mindset is a perpetuation of this trend.
What is the end result? we become dependent on the supplier; locked-in.

Yes there is vendor lock in, but you can use serverless framework and deploy to more than one cloud provider. Also if you decided to move to your own infrastructure you could use [funktion](https://funktion.fabric8.io/)

**Comment by nelsonic** [comment]
https://youtu.be/q8wueg2hswA
